### PR TITLE
annotate #[allow(unreachable_code)] to tail expr

### LINKED ISSUE #42: Unreachable expression warning in tail throw
Thanks for this crate! Minor annoyance I stumbled upon:

Given a function:

```rust
#[throws(())]
fn always_throw() {
    throw!(());
}
```

A compile warning pops out:

```
   | #[throws(())]
   | ^^^^^^^^^^^^^
   | |
   | unreachable expression
   | any code following this expression is unreachable

```

### PR DESCRIPTION
**Pull request by ubnt-intrepid**
fixes #42.

### CODE REVIEW

File: fehler-macros/src/throws.rs [hunk 1]
```diff
@@ -141,10 +141,24 @@ fn is_unit_fn(i: &syn::ReturnType) -> bool {
     }
 }
 
-fn ok(expr: &syn::Expr) -> syn::Expr {
-    syn::parse2(quote::quote!(<_ as ::fehler::__internal::_Succeed>::from_ok(#expr))).unwrap()
+fn ok(expr: &syn::Expr, is_tail: bool) -> syn::Expr {
+    if is_tail {
+        syn::parse_quote! {
+            #[allow(unreachable_code)]
+            <_ as ::fehler::__internal::_Succeed>::from_ok(#expr)
```
Comment (dtolnay) [inline fehler-macros/src/throws.rs:148]: It's not great that this ends up suppressing legitimate unreachable_code lints inside the user's tail expr (which might be a big complicated `match` or something). Is there a different way?

Comment (ubnt-intrepid) [reply fehler-macros/src/throws.rs:148]: Adding here is certainly wrong (I was confused with the case of unit return).
It seems to be desirable to add the attribute only on an implicit unit, and inherit the original attributes when an explicit tail expr is provided.

**Comment by withoutboats** [comment]
This does seem like strictly an improvement.. But I wonder if we could come up with a better way that also solves the problem in non-unit returning functions? I can't think of any, though :\

**Comment by ijackson** [comment]
Hi.  Thanks to withoutboats for this nice crate and to dtolnay for trying to improve this wart.

I found that for a function returning non-unit, the following syntax does not produce the warning:
```
#[throws(X)] fn b() -> usize { throw!(X()) ; }
```

My full test program:
```
use fehler::{throw,throws};
use thiserror::Error;

#[derive(Error,Debug)]
#[error("X")]
struct X ();

#[allow(unreachable_code)] #[throws(X)] fn a() -> usize { throw!(X())   }
                           #[throws(X)] fn b() -> usize { throw!(X()) ; }

#[allow(unreachable_code)]  #[throws(X)]    fn p()  { throw!(X())   }
#[allow(unreachable_code)]  #[throws(X)]    fn q()  { throw!(X()) ; }

#[throws(X)]
fn main() {
    a()?;
    b()?;
    p()?;
    q()?;
}
```

**Comment by ubnt-intrepid** [comment]
@ijackson IMO it is the expected behavior, since there are no statements after `throw!()`.
Imagine the case of using `panic!()` instead of `throw!()` in normal functions like `fn x() -> usize { panic!(); }`.

**Comment by withoutboats** [comment]
I think the best way to do this actually would be to change the desugaring to this:

```rust
let _x = { #body };
#[allow(unreachable_code)]
_Succeed::from_ok(_x)
```

I think this would suppress the lint properly in every case.

**Comment by ubnt-intrepid** [comment]
Close this PR since I've rewritten the patch as #46.

**Pull request closed without merging**